File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/models/Corridor.java
package com.joeshuff.dddungeongenerator.generator.models;

import android.graphics.Point;

import java.util.ArrayList;
import java.util.List;

public class Corridor {

    int startX = 0;
    int startY = 0;

    int heightX = 0; //GOES LEFT OR RIGHT
    int heightY = 0; //GOES UP OR DOWN

    public Corridor(int startX, int startY, int heightX, int heightY) {
        this.startX = startX;
        this.startY = startY;
        this.heightX = heightX;
        this.heightY = heightY;
    }

    public Corridor globalise(int startX, int startY) {
        return new Corridor(this.startX  + startX, this.startY + startY, heightX, heightY);
    }

    public static List<Corridor> toCorridors(List<Point> points) {
        return new ArrayList<>();
    }

}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/models/Corridor.kt
package com.joeshuff.dddungeongenerator.generator.models

import com.joeshuff.dddungeongenerator.db.models.Point
import com.joeshuff.dddungeongenerator.db.toRealmList
import com.joeshuff.dddungeongenerator.util.Logs
import io.realm.RealmList
import io.realm.RealmObject

open class Corridor(): RealmObject() {
    var sections: RealmList<CorridorSection> = RealmList()

    constructor(points: List<Point>): this() {
        var currentSection = CorridorSection()

        points.forEachIndexed { index, point ->
            when (index) {
                0 -> currentSection.setStart(point)
                else -> {
                    if (currentSection.isSameDirection(point)) {
                        currentSection.makeBigger()
                    } else {
                        val prev = points.getOrNull(index - 1)

                        prev?.let {
                            currentSection.finalise()
                            sections.add(currentSection)
                            currentSection = CorridorSection(it, point)
                        }?: run {
                            Logs.e("GENERATE", "previous corridor couldn't be found", Throwable("Corridor generation error"))
                        }
                    }
                }
            }
        }

        currentSection.finalise()
        sections.add(currentSection)
    }

    fun globalise(startX: Int, startY: Int): Corridor {
        val corridor = Corridor()
        corridor.sections = ArrayList(sections.map { it.globalise(startX, startY) }).toRealmList()
        return corridor
    }
}

